### PR TITLE
Implement shrink for a bunch of stuff

### PR DESCRIPTION
**Pull request by fitzgen**
More on the way, but this felt like a good checkpoint.

r? @Manishearth

### CODE REVIEW

**Comment by fitzgen** [comment]
Phew! Ok that's all of them :-p

**Comment by Manishearth** [comment]
If you're doing a lot of work on this crate can you also fix https://github.com/rust-fuzz/rust_arbitrary/issues/8 ?

File: src/lib.rs [hunk 1]
```diff
@@ -204,6 +204,14 @@ impl Arbitrary for char {
             }
         }
     }
+
+    fn shrink(&self) -> Box<dyn Iterator<Item = Self>> {
+        let x = *self as u32;
+        Box::new(x.shrink().filter_map(|x| {
+            use std::convert::TryFrom;
+            char::try_from(x).ok()
```
Comment (Manishearth) [inline src/lib.rs]: this might be problematic for unicode correctness, it may abort early for large code points

Comment (fitzgen) [reply src/lib.rs]: Hm I’m not sure how I’m familiar enough with Unicode to write this method properly. Do you have a suggested implementation or should I skip this method and drop this commit from the pull?

Comment (Manishearth) [reply src/lib.rs]: Leave it for now, just make fixing this a part of #8

**Pull request merged**